Port "truckersmp-cli" script to Python 3

kakurasan opened the pull request:
Porting `truckersmp-cli` script to Python 3, with some new features and improvements.

* Don't create/use temporary files
* Don't use external program to download files
* Long options (e.g. `--update`) can be specified
* Check MD5s when downloading mod files
* Detect Steam installation on Debian-based systems (`~/.steam/debian-installation/` or `~/.steam/`)
* Use OpenGL rendering backend by default (D3D11 can be used but it doesn't work without native 64-bit `d3dcompiler_47.dll`)
* Provide an option to activate (download + add DLL override setting) native 64-bit `d3dcompiler_47.dll` when starting (needed for D3D11 rendering backend)
* Provide an option to start singleplayer game (useful for save editing, using DXVK, etc.)
* Logging

Python ~3.2 (released in Feb 2011)~ 3.3 (released in September 2012) or later is required (for `subprocess.DEVNULL` ~and `shutil.which()`~).

List of options:

Short option|Long option|Description
---|---|---
`-a`|`--ats`|Use American Truck Simulator
`-e`|`--ets2`|Use Euro Truck Simulator 2
`-p`|`--proton`|Start the game with Proton
`-w`|`--wine`|Start the game with Wine
`-h`|`--help`|Show help
`-s`|`--start`|Start the game
`-u`|`--update`|Update the game
`-v`|`--verbose`|Verbose output (none:error, once:info, twice or more:debug)
`-g DIR`|`--gamedir DIR`|Choose a different directory for the game files [Default: `$XDG_DATA_HOME/truckersmp-cli/(Game name)/data`]
`-i APPID`|`--proton-appid APPID`|Choose a different AppId for Proton (Needs an update for changes)
`-m DIR`|`--moddir DIR`|Choose a different directory for the mod files [Default: `$XDG_DATA_HOME/truckersmp-cli/TruckersMP`, Fallback: `./truckersmp`]
`-n NAME`|`--account NAME`|Steam account name to use
`-o DIR`|`--protondir DIR`|Choose a different Proton directory [Default: $XDG_DATA_HOME/truckersmp-cli/Proton]
`-l LOG`|`--logfile LOG`|Write log into LOG, `-vv` option is recommended [Default: Empty string (only stderr)] Note: Messages from Steam/steamcmd won't be written, only from this script (Game logs are written into `My Documents/{ETS2,ATS}MP/logs/client_*.log`)
`-x DIR`|`--prefixdir DIR`|Choose a different directory for the prefix [Default: `$XDG_DATA_HOME/truckersmp-cli/(Game name)/prefix`]
`-c`|`--activate-native-d3dcompiler-47`|Activate native 64-bit `d3dcompiler_47.dll` when starting (Needed for D3D11 renderer)
(Not available)|`--use-wined3d`|Use OpenGL-based D3D11 instead of DXVK when using Proton
(Not available)|`--enable-d3d11`|Use Direct3D 11 instead of OpenGL
(Not available)|`--disable-proton-overlay`|Disable Steam Overlay when using Proton
(Not available)|`--beta VERSION`|Set game version to VERSION, useful for downgrading (e.g. `temporary_1_35`)
(Not available)|`--singleplayer`|Start singleplayer game, useful for save editing, using/testing DXVK in singleplayer, etc.)

lhark commented on truckersmp-cli:
```diff
@@ -1,512 +1,547 @@
+            res = conn.getresponse()
+
+            if res.status != 200:
+                printerr("Server response code is not 200.")
```
Printing the actual status code would be more useful. Something like `"Error: server {} responded with status code {}".format(host, res.status)`

lhark commented on truckersmp-cli:
```diff
@@ -1,512 +1,547 @@
+
+            if md5hash.hexdigest() != md5:
+                print("")  # newline
+                printerr("MD5 mismatch for {}".format(dest))
```
Is it intended that the newline is written to stdout and the error message to stderr. If it's simply to let the error message breathe, send the newline to stderr too.

kakurasan replied:
Not intended to print a blank line. The `print()` for download progress doesn't print a newline:
```py
print("\r{:40}{:>40}".format(name, progress), end="")
```
Now, if MD5 doesn't match, the script prints "[filename]   MD5 ERROR" with a newline:
```py
if md5hash.hexdigest() != md5:
    print("\r{:40}{:>40}".format(name, "MD5 ERROR"))
    printerr("MD5 mismatch for {}".format(dest))
    return False
```

lhark commented on truckersmp-cli:
```diff
@@ -1,512 +1,547 @@
+    try:
+        with open("/usr/games/steam") as f:
+            for l in f:
+                if l.rstrip() == "config=$HOME/.steam":
```
Is `/usr/games/steam` exclusive to debian ? It is not present on arch, but maybe other systems do something similar and instead of testing for `l == "config=$HOME/.steam"`, we could get the actual value from the `config=` field.

Lucki replied:
Does Debian really use `~/.steam`? That's the old directory which is still there as fallback. There are only a few legacy files and symlinks to `$XDG_DATA_HOME/Steam` nowadays.

~~~
$ cat /usr/lib/steam/steam

# Get the default data path
	STEAM_DATA_HOME="${XDG_DATA_HOME:-$HOME/.local/share}"
	case "$STEAMPACKAGE" in
		steam)
			CLASSICSTEAMDIR="$HOME/Steam"
			DEFAULTSTEAMDIR="$STEAM_DATA_HOME/Steam"
~~~

lhark replied:
I asked a friend on Ubuntu how Steam was setup on her machine : `~/.steam` was an actual folder, not a symlink. `~/.local/share/steam` also existed, but was not used by steam.
That's anecdotal data, but it mean that a setup we might encounter.

kakurasan replied:
Though `$XDG_DATA_HOME/Steam` is not used on my Ubuntu 19.04, I removed the "detection" code for now. See #32. If we can add an option to specify the installation directory (such as `--steamdir DIR`), I'll add it.

kakurasan replied:
Now the script detects timestamp update in one of:
* `$XDG_DATA_HOME/Steam/config/loginusers.vdf`
* `~/.steam/config/loginusers.vdf`
* `~/.steam/debian-installation/config/loginusers.vdf`
and `inotifywait` is not used.

Lucki replied:
But the maximum wait time is now 30 seconds? Steam takes longer for me to boot up completely and log me in.

kakurasan replied:
Changed to 99 seconds. I hope this is enough.

lhark commented on truckersmp-cli:
```diff
@@ -1,512 +1,547 @@
-cat <<WINE_USAGE
-
+            printerr("waiting 30 seconds for steam to start up.")
+            time.sleep(30)
```
If we want to be fancy, let's add a countdown here. That's just an idea though, not required for me to accept this PR ^^

lhark commented on truckersmp-cli:
```diff
@@ -1,512 +1,547 @@
+        else:
+            raise Exception
+    except:
+        printdbg("Better do not self update")
```
Shouldn't it be `"Better not to self update"` ?

lhark commented on truckersmp-cli:
```diff
@@ -1,512 +1,547 @@
+        else:
+            printdbg("Found running steam, running pkill -x steam")
+            subprocess.call(["pkill", "-x", "steam"])
+            time.sleep(5)
```
Maybe we could do better than just waiting. I think we can use `os.kill` directly in a loop to test for process termination (os.waitpid only work with child processes). Also we could get rid of the subprocess call.

cf. https://stackoverflow.com/a/7654102

lhark commented on truckersmp-cli:
```diff
@@ -1,512 +1,547 @@
+      "+app_update", args.steamid, "validate",
+      "+quit"])
+
+def check_error():
```
Maybe `check_args_errors()` would be clearer ?

lhark commented on truckersmp-cli:
```diff
@@ -1,512 +1,547 @@
-        fi
-    fi
+    if not args.update and not args.start:
+        sys.exit("What to do? Starting (-s) or updating (-u) the game?")
```
I feel like if the script receives zero arguments, we should show the usage string instead of this message.

Lucki commented on truckersmp-cli:
```diff
@@ -1,512 +1,547 @@
+        "WINEPREFIX" : args.prefixdir,
+    }
+    for key, value in env.items():
+        os.environ[key] = value
```
Looks like you're overwriting the users environment here. Maybe it's a good idea to temporary save them and reset them afterwards to the old values.

aearil commented on truckersmp-cli:
```diff
@@ -1,64 +1,81 @@
+def printdbg(msg, end="\n"):
     if args.verbose:
-        print(msg, end = end)
+        print(msg, end=end)
```
printdbg should also send the messages to stderr.

stdout should only be used for the actual output of the program, the kind of output that would be then parsed by a subsequent pipe.

aearil commented on truckersmp-cli:
```diff
@@ -292,55 +315,67 @@ Please report an issue: {}""".format(e, URL.issueurl))
+                    os.kill(int(pid), signal.SIGTERM)
+            except OSError as e:
+                if e.errno == errno.ESRCH:
+                    process_count -= 1
```
If another user has an instance of steam open, this will loop infinitely. We could either check for `errno.EPERM` or use `pgrep -u $USER -x steam` to limit the PID search to those owned by the user.

aearil commented:
I just discovered this repo, thank you so much for this handy tool.
I hope my impromptu review was not out of place, I just want to help :)

Lucki commented on truckersmp-cli:
```diff
@@ -1,512 +1,596 @@
+        while len(steam_pid_list) > 0:
+            pid = int(steam_pid_list[0])
+            try:
+                os.kill(pid, signal.SIGTERM)
```
I just found out about `steam -shutdown` maybe that's more suitable to quit the process.

squadramunter replied:
Does this include a fix for this? https://github.com/lhark/truckersmp-cli/issues/34

kakurasan replied:
This PR is not intended to fix it (but this PR fixes #32).

parkerlreed commented:
What do I need to provide to debug this? I'm running into Unable to find ETS2 or ATS in this GAMEDIR and have come here as per suggestion.

Installation runs fine but when I go to launch it, it still seems to error out.

Arch Linux
Main Steam folder and Proton are on main SSD, game library on external HDD

![image](https://user-images.githubusercontent.com/841440/69385970-1fb47c00-0c8f-11ea-81ef-131d2aac00e6.png)

![image](https://user-images.githubusercontent.com/841440/69386061-75892400-0c8f-11ea-91d3-5128dca304e0.png)

```
$ ./truckersmp-cli -eusp -n parkerlreed
...
...
Waiting for user info...OK
 Update state (0x3) reconfiguring, progress: 0.00 (0 / 0)
 Update state (0x5) validating, progress: 15.94 (574103108 / 3601778063)
 Update state (0x5) validating, progress: 34.46 (1240997444 / 3601778063)
 Update state (0x5) validating, progress: 52.80 (1901600324 / 3601778063)
 Update state (0x5) validating, progress: 71.14 (2562203204 / 3601778063)
 Update state (0x5) validating, progress: 89.71 (3231194692 / 3601778063)
Success! App '227300' fully installed.
Traceback (most recent call last):
  File "/home/parker/.local/share/truckersmp-cli/Proton/proton", line 726, in <module>
    g_session.init_session()
  File "/home/parker/.local/share/truckersmp-cli/Proton/proton", line 496, in init_session
    g_compatdata.setup_prefix()
  File "/home/parker/.local/share/truckersmp-cli/Proton/proton", line 296, in setup_prefix
    with self.prefix_lock:
  File "/home/parker/.local/share/truckersmp-cli/Proton/filelock.py", line 323, in __enter__
    self.acquire()
  File "/home/parker/.local/share/truckersmp-cli/Proton/filelock.py", line 271, in acquire
    self._acquire()
  File "/home/parker/.local/share/truckersmp-cli/Proton/filelock.py", line 384, in _acquire
    fd = os.open(self._lock_file, open_mode)
FileNotFoundError: [Errno 2] No such file or directory: '/home/parker/.local/share/truckersmp-cli/Euro Truck Simulator 2/prefix/pfx.lock'
```

kakurasan commented:
You might want to try `--gamedir DIR`(`-g DIR`) option. `DIR` is your game directory.
If you want to debug, insert the code: `printdbg("Game directory:", args.gamedir)`

parkerlreed commented:
Does gamedir want the Steam installed dir or the steamcmd download run by the script? I would assume it would be using what it just downloaded.

When I checked that folder it didn't seem to create the prefix. Which would lead to that error above. Before testing and posting that irst comment I nuked the entire ~/.local/share/truckerms-cli folder so I started fresh with only this python.

```
[parker@e55 Euro Truck Simulator 2]$ pwd
/home/parker/.local/share/truckersmp-cli/Euro Truck Simulator 2
[parker@e55 Euro Truck Simulator 2]$ ls
data
```

kakurasan commented:
@parkerlreed Sorry. I found/reproduced the issue (The Python3 script doesn't create new prefix directory when it doesn't exist) and fixed it. Could you retest with latest?

parkerlreed commented:
Just got back around to trying this. Appears to launch fine and the menu comes up but mouse interaction is doing nothing. I can get partially through the menus with keyboard.

Is there a way to tell what Proton version this is picking? I have a few custom builds in compatibilitytools.d and I'm not sure if they are interfering.

parkerlreed commented:
Testing again I can see that I have control of the mouse for about 1 second after it loads but then it just comes to a stop.

https://www.youtube.com/watch?v=QAO4Gukd-wk

kakurasan commented:
For now, Windows version of `d3dcompiler_47.dll` (64-bit) is needed for D3D11 renderer. It can be installed via [Winetricks](https://github.com/Winetricks/winetricks). Built-in (Wine/Proton) version of `d3dcompiler_47.dll` is incomplete and not working (It outputs a lot of `<error> Failed to load texture 'xxx.png'.` messages and doesn't show TruckersMP login screen).

```
(Show Proton version)
$ cat ~/.local/share/truckersmp-cli/Proton/version
1574789123 proton-4.11-9

(Install d3dcompiler_47 / ETS2 default prefix)
$ WINEPREFIX=~/.local/share/truckersmp-cli/Euro\ Truck\ Simulator\ 2/prefix/pfx WINE=~/.local/share/truckersmp-cli/Proton/dist/bin/wine winetricks d3dcompiler_47

(Install d3dcompiler_47 / ATS default prefix)
$ WINEPREFIX=~/.local/share/truckersmp-cli/American\ Truck\ Simulator/prefix/pfx WINE=~/.local/share/truckersmp-cli/Proton/dist/bin/wine winetricks d3dcompiler_47
```

IMPORTANT: If you want to use DXVK, [version 1.4.6](https://github.com/doitsujin/dxvk/releases/tag/v1.4.6) or newer is needed (See https://github.com/doitsujin/dxvk/issues/1250). I needed to replace existing DXVK DLLs (`/path/to/Proton/dist/lib64/wine/dxvk/`) with v1.4.6 DLLs as of Proton 4.11-9.

I tested "Proton 4.11-9 + DXVK 1.4.6 + ETS2MP + native `d3dcompiler_47.dll`" and it's working.

parkerlreed commented:
Thanks for the updates. Any idea why the mouse is killing itself? Same result without the --use-wined3d

```
[parker@e55 ~]$ cat ~/.local/share/truckersmp-cli/Proton/version
1574789123 proton-4.11-9
```

kakurasan commented:
> Any idea why the mouse is killing itself?

I guess D3D11 renderer is being used without native `d3dcompiler_47.dll` on your system. Does your terminal show `<error> Failed to load texture 'xxx.png'.` messages? You can check `~/.local/share/truckersmp-cli/*/prefix/pfx/drive_c/users/steamuser/My Documents/ETS2MP/logs/client_xx_xx_xxxx.log`.

Recently (ETS2/ATS v1.36) D3D9 support has been dropped and this is a major change for us. There are 2 options:

* Use D3D11 renderer with native `d3dcompiler_47.dll`: When built-in (Wine/Proton) version of `d3dcompiler_47.dll` is used, TruckersMP login screen is not shown and mouse cursor doesn't move. When Windows version of `d3dcompiler_47.dll` (64-bit) is used, TruckersMP login screen is shown. I added `--activate-native-d3dcompiler-47`(`-c`) option that downloads/activates the Windows version of DLL (Winetricks is not needed). If you want to use D3D11 renderer, use this option with `--start`(`-s`). If you want to use DXVK, use version 1.4.6 or newer.
* Use OpenGL renderer: When D3D11 is not available, OpenGL renderer is used as fallback. I added `--disable-d3d11` option that disables D3D11. If you want to use OpenGL renderer, use this option with `--start`(`-s`). See also https://github.com/lhark/truckersmp-cli/issues/42

parkerlreed commented:
Thank you. Will check this out when I get a chance. Appreciate the updates!

parkerlreed commented:
Well the native d3dcompiler is definitely out of the question for me... :D https://youtu.be/MllG-N2tj_o?t=19

I'll try and get the DXVK updated

parkerlreed commented:
What arguments do I need for DXVK?

kakurasan commented:
> Oh it auto used DXVK after the native compiler install

DXVK is used by default when using Proton. wined3d can be used if `PROTON_USE_WINED3D=1`(`--use-wined3d` option in this PR) is specified, but I don't recommend it because OpenGL renderer is faster than wined3d.

> https://youtu.be/MllG-N2tj_o?t=19

DXVK < 1.4.6 was used when the video was captured, right?

parkerlreed commented:
I assume so. It cleared up after 1.4.6. I just thought it was funny.

kakurasan commented:
> I assume so. It cleared up after 1.4.6. I just thought it was funny.

All right, thank you for testing and comments.

Have fun!

Lucki commented:
Haven't seen it by looking at the code, do you mind to include #39?

kakurasan commented:
It has been implemented. This is enabled by default and can be disabled by `--disable-proton-overlay` option.
Frame rate can be displayed with this "Steam Overlay" and/or "D3D11 + DXVK + environment variable `DXVK_HUD`" when using Proton.

kakurasan commented:
@parkerlreed @Lucki *OpenGL* renderer is used by default now because it was less user-friendly (D3D11 has `d3dcompiler_47.dll` issue). If you want to use D3D11, use `--enable-d3d11` option.

Klairm commented:
I placed mods on ~/.local/share/truckersmp-cli/TruckersMP/data/ets2/mods but still not detecting the mods, what can i do?

kakurasan commented:
@Klairm
"Multiplayer mod" files are downloaded and verified automatically and you don't need to place them manually.

If you see `<error> Failed to load texture 'xxx.png'.` messages and login screen is not shown, try OpenGL renderer (don't use `--enable-d3d11` option) or `--activate-native-d3dcompiler-47` option (to use native `d3dcompiler_47.dll`).

If you mean "Winter mod" and you're using Proton(`-p`), `*.scs` files need to be placed into `~/.local/share/truckersmp-cli/Euro Truck Simulator 2/prefix/pfx/drive_c/users/steamuser/My Documents/ETS2MP/mod/`.

Klairm commented:
> @Klairm
> "Multiplayer mod" files are downloaded and verified automatically and you don't need to place them manually.
> 
> If you see `<error> Failed to load texture 'xxx.png'.` messages and login screen is not shown, try OpenGL renderer (don't use `--enable-d3d11` option) or `--activate-native-d3dcompiler-47` option (to use native `d3dcompiler_47.dll`).
> 
> If you mean "Winter mod" and you're using Proton(`-p`), `*.scs` files need to be placed into `~/.local/share/truckersmp-cli/Euro Truck Simulator 2/prefix/pfx/drive_c/users/steamuser/My Documents/ETS2MP/mod/`.

I mean any custom mod, I tried the path you said and still not detecting them, strange

kakurasan commented:
@Klairm Default paths to mod directories (for Proton) are:

* ETS2: `$XDG_DATA_HOME/truckersmp-cli/Euro Truck Simulator 2/prefix/pfx/drive_c/users/steamuser/My Documents/Euro Truck Simulator 2/mod/`
* ATS: `$XDG_DATA_HOME/truckersmp-cli/American Truck Simulator/prefix/pfx/drive_c/users/steamuser/My Documents/American Truck Simulator/mod/`

Klairm commented:
> @Klairm Default paths to mod directories (for Proton) are:
> 
>     * ETS2: `$XDG_DATA_HOME/truckersmp-cli/Euro Truck Simulator 2/prefix/pfx/drive_c/users/steamuser/My Documents/Euro Truck Simulator 2/mod/`
> 
>     * ATS: `$XDG_DATA_HOME/truckersmp-cli/American Truck Simulator/prefix/pfx/drive_c/users/steamuser/My Documents/American Truck Simulator/mod/`

Also not working, I tried a lot of paths for the mod, strange issue, and if I use the -m option for a custom path it will also not detect it

kakurasan commented:
@Klairm
The [paths](https://github.com/lhark/truckersmp-cli/pull/31#issuecomment-573668611) work for me in singleplayer game (`--singleplayer` option is needed).
ETS2:
![singleplayer-ets2-mods-detected](https://user-images.githubusercontent.com/10556453/72349710-61f2ba80-3720-11ea-8b43-86148711e121.png)
ATS:
![singleplayer-ats-mods-detected](https://user-images.githubusercontent.com/10556453/72349718-661ed800-3720-11ea-9879-bb79949728ba.png)

AFAIK almost all 3rd party mods are not compatible with TruckersMP. (ProMods is supported but I can't test "TruckersMP + ProMods" because I don't have map DLCs.)

`-m` option changes the directory for "Multiplayer mod" and it's unrelated to your issue.

If the paths don't work, more information is needed:

* Mod information
* Command line and terminal output (use `--verbose` option)

Klairm commented:
> @Klairm
> The [paths](https://github.com/lhark/truckersmp-cli/pull/31#issuecomment-573668611) work for me in singleplayer game (`--singleplayer` option is needed).
> ETS2:
> ![singleplayer-ets2-mods-detected](https://user-images.githubusercontent.com/10556453/72349710-61f2ba80-3720-11ea-8b43-86148711e121.png)
> ATS:
> ![singleplayer-ats-mods-detected](https://user-images.githubusercontent.com/10556453/72349718-661ed800-3720-11ea-9879-bb79949728ba.png)
> 
> AFAIK almost all 3rd party mods are not compatible with TruckersMP. (ProMods is supported but I can't test "TruckersMP + ProMods" because I don't have map DLCs.)
> 
> `-m` option changes the directory for "Multiplayer mod" and it's unrelated to your issue.
> 
> If the paths don't work, more information is needed:
> 
>     * Mod information
> 
>     * Command line and terminal output (use `--verbose` option)

The mod is the Speed Limit Indicator ( downloaded here https://trucksimulator.org/ets2-speed-limit-indicator-v1-0-1-35-x/)

And there's the information using `--verbose` ( I don't see anything useful here)  ->  https://pastebin.com/2G0yGGrZ

kakurasan commented:
> Startup command:
>   ...
>   /usr/bin/python3 /home/ivan/.local/share/truckersmp-cli/Proton/proton
>   run
>   /home/ivan/TruckersMP/python-mp/truckersmp-cli/truckersmp-cli.exe /home/ivan/.local/share/truckersmp-cli/Euro Truck Simulator 2/data /home/ivan/.local/share/truckersmp-cli/TruckersMP

@Klairm You were trying to use unsupported mod in *multiplayer* game (Proton ran `truckersmp-cli.exe`: This means `--singleplayer` option was not used).

Please note that *almost all 3rd party mods are not compatible with TruckersMP* (AFAIK only Winter mod and ProMods are allowed).

The mod works for me in singleplayer game:
![singleplayer-ets2-speed_limit_indicator](https://user-images.githubusercontent.com/10556453/72435294-16ecac00-37e1-11ea-80d0-be7115d548f2.png)

It cannot be used in multiplayer game (but this behavior is *expected*):
![multiplayer-ets2-mods-not-detected](https://user-images.githubusercontent.com/10556453/72435304-1e13ba00-37e1-11ea-8995-9c72ef767ca0.png)

If you want to use the mod in singleplayer game, you need to use `--singleplayer` option.

Klairm commented:
> > Startup command:
> > ...
> > /usr/bin/python3 /home/ivan/.local/share/truckersmp-cli/Proton/proton
> > run
> > /home/ivan/TruckersMP/python-mp/truckersmp-cli/truckersmp-cli.exe /home/ivan/.local/share/truckersmp-cli/Euro Truck Simulator 2/data /home/ivan/.local/share/truckersmp-cli/TruckersMP
> 
> @Klairm You were trying to use unsupported mod in _multiplayer_ game (Proton ran `truckersmp-cli.exe`: This means `--singleplayer` option was not used).
> 
> Please note that _almost all 3rd party mods are not compatible with TruckersMP_ (AFAIK only Winter mod and ProMods are allowed).
> 
> The mod works for me in singleplayer game:
> ![singleplayer-ets2-speed_limit_indicator](https://user-images.githubusercontent.com/10556453/72435294-16ecac00-37e1-11ea-80d0-be7115d548f2.png)
> 
> It cannot be used in multiplayer game (but this behavior is _expected_):
> ![multiplayer-ets2-mods-not-detected](https://user-images.githubusercontent.com/10556453/72435304-1e13ba00-37e1-11ea-8995-9c72ef767ca0.png)
> 
> If you want to use the mod in singleplayer game, you need to use `--singleplayer` option.

Okay thx, is there any list of supported mods on truckersmp or just promods and winter mod are allowed?

Lucki commented:
> is there any list of supported mods

It looks like it needs save editing to use mods online. While doing this they have to [comply to the rules](https://truckersmp.com/rules#3-save-editing-rules).

Lucki commented:
@lhark what's holding this back? Would you mind to add some more active maintainers?

Lucki commented:
Today I got a `302` from `download.ets2.com` which python doesn't seem to follow. Investigating on Firefox it resolves to `https://failover.truckersmp.com/`. `downloads.ets2.com` gave me `502`.
We should probably add this as fallback or follow the redirect.

~~~
** ERROR **  Server download.ets2mp.com responded with status code 302.
** ERROR **  Server downloads.ets2mp.com responded with status code 502.
~~~

Lucki commented on truckersmp-cli:
```diff
@@ -1,512 +1,770 @@
-    fi
-    print "Steamcmd: $steamcmd"
+    logging.debug("Closing Steam")
+    subproc.call(["steam", "-shutdown"])
```
Just found out that this will start steam if it isn't already open and will just close it again. Maybe a runtime check would prevent this.

lhark commented:
@Lucki I've given you and @kakurasan access to the repo, which I should have done at least a year ago. I'm sorry I've left this repo hanging for so long :( 

I want to thank you for the work you put into improving my little hack. I hope that now with full access to the repo you can continue to push linux gaming forward :)

kakurasan commented:
@Lucki I'd like to combine the commits into a single commit before merging. Is it OK?

Lucki commented:
Yes, polish it as you like. We need this in master since master is broken quite some time already.

kakurasan commented:
@Lucki Ready to go. I won't merge this myself because this is my PR. Could you merge this pull request instead of me? You can see/click "Merge pull request" button in this page. Thanks.

kakurasan commented:
Thank you for all the feedbacks, comments, and merging (@Lucki).
